File: sort/c/shellsort.c
#include "algorithms.h"

#define GAP 2

void shellsort(int *array, size_t length) {
  size_t h = length / GAP;

  while (h > 0) {
    for (size_t i = h; i < length; i++) {
      for (size_t j = i; j > h - 1 && array[j] < array[j - h]; j -= h) {
        swap(&array[j], &array[j - h]);
      }
    }
    h /= GAP;
  }
}

File: sort/java/ShellSort.java
public class ShellSort {
    private static final int GAP = 2;

    public static void sort(int[] array) {
        int h = array.length / GAP;
        while (h > 0) {
            for (int i = h; i < array.length; i++) {
                for (int j = i; j >= h && array[j] < array[j - h]; j -= h) {
                    Utils.swap(array, j, j - h);
                }
            }
            h /= GAP;
        }
    }
}
